File: p1.py
#Tc = o(n)
#SC = O(n)
#successfully submitted on leetcode; yes

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = [0 for i in range(len(nums))]
        right = [0 for i in range(len(nums))]

        #left pass
        left[0] = 1
        for i in range(1, len(nums)):
            left[i] = nums[i-1] * left[i-1]
        
        #right pass
        right[-1] = 1
        for i in range(len(nums)-2, -1, -1):
            right[i] = nums[i+1] * right[i+1]
        
        ans = []
        for i in range(0, len(nums)):
            ans.append(left[i]*right[i])
        return ans 

File: p1.java
// Time Complexity : O(n)
// Space Complexity : Constant O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
//Take running sum from left and store it in the result array. Take running sum from the right and store it in the running sum variable. Multiple the left running sum and right running sum to get the answer

// Your code here along with comments explaining your approach

class Solution {
    public int[] productExceptSelf(int[] nums) {

        // Result array to store temp result and then final result
        int[] result = new int[nums.length];
        // Storing first element as 1 as there is nothin to the left
        result[0] = 1;
        // Traversing through the array to get left ruuning sum
        for (int i = 1; i < nums.length; i++) {
            result[i] = result[i - 1] * nums[i - 1];
        }
        // Traversing through the array from right to get the running sum
        // Multiplying it with left running sum to get the result
        int rs = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            result[i] = result[i] * rs;
            rs = rs * nums[i];
        }

        return result;
    }
}